### PR TITLE
feat: add support for enforcement of approval policy in deployment and configuration change

### PR DESCRIPTION
**Pull request by eshankvaish**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes https://github.com/devtron-labs/sprint-tasks/issues/1131

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

# Checklist:

* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/app/types.ts
Comment (AbhishekA1509) [inline src/components/app/types.ts:106]: Please check if its feasible to have this key in app details since we poll it and Can sync with prakhar since he was working on block state api

File: src/Pages/Shared/EnvironmentOverride/EnvironmentOverride.tsx
Comment (AbhishekA1509) [inline src/Pages/Shared/EnvironmentOverride/EnvironmentOverride.tsx:65]: Can re-write as appOrEnvIdToAppOrEnvApprovalConfigMap[+params[isAppGroupView ? 'appId' : 'envId']]

File: src/Pages/Shared/EnvironmentOverride/EnvironmentOverride.tsx
Comment (AbhishekA1509) [inline src/Pages/Shared/EnvironmentOverride/EnvironmentOverride.tsx:163]: Suggestion: Can abstract further, plus if doing here please look into release config as well

File: src/Pages/Applications/DevtronApps/Details/AppConfigurations/AppConfig.types.ts
Comment (AbhishekA1509) [inline src/Pages/Applications/DevtronApps/Details/AppConfigurations/AppConfig.types.ts:47]: Please remove

File: src/Pages/Applications/DevtronApps/Details/AppConfigurations/Navigation/Navigation.helper.tsx
Comment (AbhishekA1509) [inline src/Pages/Applications/DevtronApps/Details/AppConfigurations/Navigation/Navigation.helper.tsx:29]: We don't need to check !isLocked here

File: src/Pages/Applications/DevtronApps/Details/AppConfigurations/Navigation/EnvironmentOverrideRouter.tsx
Comment (AbhishekA1509) [inline src/Pages/Applications/DevtronApps/Details/AppConfigurations/Navigation/EnvironmentOverrideRouter.tsx:327]: Pls move to const

File: src/Pages/Applications/DevtronApps/Details/AppConfigurations/MainContent/DeploymentConfigCompare/DeploymentConfigCompare.tsx
Comment (AbhishekA1509) [inline src/Pages/Applications/DevtronApps/Details/AppConfigurations/MainContent/DeploymentConfigCompare/DeploymentConfigCompare.tsx:35]: Pls remove

File: src/Pages/Applications/DevtronApps/Details/AppConfigurations/MainContent/DeploymentConfigCompare/utils.ts
Comment (AbhishekA1509) [inline src/Pages/Applications/DevtronApps/Details/AppConfigurations/MainContent/DeploymentConfigCompare/utils.ts:220]: Suggestion: Can keep this in fe-lib

File: src/Pages/Shared/ConfigMapSecret/ConfigMapSecretProtected.tsx
Comment (AbhishekA1509) [inline src/Pages/Shared/ConfigMapSecret/ConfigMapSecretProtected.tsx:188]: Suggestion: Can make it a component in fe-lib since re-using

File: src/Pages/Shared/ConfigMapSecret/ConfigMapSecretProtected.tsx
Comment (AbhishekA1509) [inline src/Pages/Shared/ConfigMapSecret/ConfigMapSecretProtected.tsx:186]: TODO: In case of patch strategy we also show an approve button in dry run view pls look into that as well

File: src/Pages/Shared/ConfigMapSecret/ConfigMapSecretProtected.tsx
Comment (AbhishekA1509) [inline src/Pages/Shared/ConfigMapSecret/ConfigMapSecretProtected.tsx:185]: Query: Can i edit the draft after approving, if yes, will my approval be valid?

File: src/components/app/types.ts
Comment (eshankvaish) [reply src/components/app/types.ts:106]: Pending on Gireesh to confirm this

File: src/components/app/details/triggerView/TriggerView.tsx
Comment (eshankvaish) [reply src/components/app/details/triggerView/TriggerView.tsx:766]: Will pick this later on if time permits

File: src/components/app/details/appDetails/SourceInfo.tsx
Comment (eshankvaish) [reply src/components/app/details/appDetails/SourceInfo.tsx:126]: Not required imo

File: src/components/ApplicationGroup/Details/TriggerView/EnvTriggerView.tsx
Comment (eshankvaish) [reply src/components/ApplicationGroup/Details/TriggerView/EnvTriggerView.tsx:1400]: Can be done later on if time permits

File: src/Pages/Shared/ConfigMapSecret/ConfigMapSecretProtected.tsx
Comment (eshankvaish) [reply src/Pages/Shared/ConfigMapSecret/ConfigMapSecretProtected.tsx:186]: Sure will do after the story is merged

File: src/Pages/Shared/ConfigMapSecret/ConfigMapSecretProtected.tsx
Comment (eshankvaish) [reply src/Pages/Shared/ConfigMapSecret/ConfigMapSecretProtected.tsx:185]: You have already discussed ig

File: src/Pages/Shared/ConfigMapSecret/ConfigMapSecretProtected.tsx
Comment (AbhishekA1509) [reply src/Pages/Shared/ConfigMapSecret/ConfigMapSecretProtected.tsx:188]: I believe we should avoid **duplication** wherever possible

**Review by AbhishekA1509 — APPROVED** [review]

File: src/Pages/Applications/DevtronApps/Details/AppConfigurations/appConfig.scss
Comment (vivek-devtron) [inline src/Pages/Applications/DevtronApps/Details/AppConfigurations/appConfig.scss:1]: We should remove config-protection__side-nav CSS

**Review by vivek-devtron — APPROVED** [review]

**Review by vivek-devtron — APPROVED** [review]

**Pull request merged**
